api/samples/bbv.cpp

```diff
@@ -0,0 +1,342 @@
+        // Add the newly formed BBV to the list of BBVs.
+        drvector_append(&bbvs, bbv);
+
+        // Clear hit_count_table setting all the hit counts to zero.
+        hashtable_apply_to_all_payloads(&hit_count_table, set_count_to_zero);
+    }
+}
+
+static void
+event_exit(void)
+{
+    // Get the current working directory where drrun is executing. We save the .bb file
+    // there.
+    char cwd[MAXIMUM_PATH];
+    if (!dr_get_current_directory(cwd, sizeof(cwd)))
+        FATAL("ERROR: dr_get_current_directory() failed");
+
+    // Create and open the .bb file.
+    char buf[MAXIMUM_PATH];
+    file_t bbvs_file =
+        drx_open_unique_appid_file(cwd, dr_get_process_id(), "proc", "bb",
+                                   DR_FILE_WRITE_REQUIRE_NEW, buf, sizeof(buf));
+    if (bbvs_file == INVALID_FILE)
+        FATAL("ERROR: unable to create BBVs file");
+
+    // Define the format strings we use to write the file.
+    const char *one_pair_only = "T:%d:%d \n";
+    const char *first_pair = "T:%d:%d ";
+    const char *middle_pair = ":%d:%d ";
+    const char *last_pair = ":%d:%d \n";
+
+    for (uint i = 0; i < bbvs.entries; ++i) {
+        drvector_t *bbv = (drvector_t *)drvector_get_entry(&bbvs, i);
+        for (uint j = 0; j < bbv->entries; ++j) {
+            bb_id_count_pair_t *pair = (bb_id_count_pair_t *)drvector_get_entry(bbv, j);
+
+            const char *format_string;
+            // Only one BB and related hit count for this BBV.
+            if (j == 0 && j == bbv->entries - 1) {
+                format_string = one_pair_only;
+            } else if (j == 0) {
+                format_string = first_pair;
+            } else if (j == bbv->entries - 1) {
+                format_string = last_pair;
+            } else {
+                format_string = middle_pair;
+            }
+
+            char msg[64];
+            int len = dr_snprintf(msg, sizeof(msg) / sizeof(msg[0]), format_string,
+                                  pair->id, pair->count);
+            DR_ASSERT(len > 0);
+
+            dr_write_file(bbvs_file, msg, (size_t)len);
+        }
+    }
+
+    dr_close_file(bbvs_file);
+
+    // Free DR memory.
+    hashtable_delete(&pc_to_id_map);
+    hashtable_delete(&bb_size_table);
+    hashtable_delete(&hit_count_table);
+    if (!drvector_delete(&bbvs))
+        FATAL("ERROR: BBVs drvector not deleted");
+
+    drx_exit();
+    drmgr_exit();
+
+    // Avoid accumulation of option values on static-link re-attach.
+    droption::droption_parser_t::clear_values();
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
+                      bool for_trace, bool translating, void *user_data)
+{
+    // By default drmgr enables auto-predication, which predicates all instructions
+    // with the predicate of the current instruction on ARM. We disable it here
+    // because we want to unconditionally execute the following instrumentation.
```

Comment:
So you'll include unexecuted predicated instructions in the bb size count?  Worth a comment.

Reply:
Done.
Yes, from my testing it seems that `perf` will count predicated instructions with a false predicate as well (probably as a no-op), so we want to include them to match the perf instruction count (necessary when we validate the resulting representative intervals).